pulsar-io/azure-data-explorer/src/main/java/org/apache/pulsar/io/azuredataexplorer/ADXSinkConfig.java

```diff
@@ -0,0 +1,105 @@
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.Map;
+import java.util.Objects;
+import lombok.Data;
+import lombok.experimental.Accessors;
+import org.apache.pulsar.io.core.annotations.FieldDoc;
+
+
+@Data
+@Accessors(chain = true)
+public class ADXSinkConfig implements Serializable {
+
+    @FieldDoc(required = true, defaultValue = "", help = "The ADX cluster URL")
+    private String clusterUrl;
+
+    @FieldDoc(required = true, defaultValue = "", help = "The database name to which data need to be ingested")
+    private String database;
+
+    @FieldDoc(required = true, defaultValue = "", help = "Table name to which pulsar data need to be ingested.")
+    private String table;
+
+    @FieldDoc(defaultValue = "", help = "The AAD app Id for authentication")
+    private String appId;
+
+    @FieldDoc(defaultValue = "", help = "The AAD app secret for authentication")
+    private String appKey;
+
+    @FieldDoc(defaultValue = "", help = "The tenant Id for authentication")
+    private String tenantId;
+
+    @FieldDoc(defaultValue = "", help = "The Managed Identity credential for authentication."
+            + " Set this with clientId in case of User assigned MI."
+            + " and 'system' in case of System assigned managed identity")
+    private String managedIdentityId;
+
+    @FieldDoc(defaultValue = "", help = "The mapping reference for ingestion")
+    private String mappingRefName;
+
+    @FieldDoc(defaultValue = "CSV", help = "The type of mapping reference provided")
+    private String mappingRefType;
+
+    @FieldDoc(defaultValue = "false", help = "Denotes if flush should happen immediately without aggregation. "
+            + "Not recommended to enable flushImmediately for production workloads")
+    private boolean flushImmediately = false;
+
+    @FieldDoc(defaultValue = "100", help = "For batching, this defines the number of "
+            + "records to hold for batching, to sink data to adx")
+    private int batchSize = 100;
+
+    @FieldDoc(defaultValue = "10000", help = "For batching, this defines the time to hold"
+            + " records before sink to adx")
+    private long batchTimeMs = 10000;
+
+    @FieldDoc(defaultValue = "1", help = "Max retry attempts, In case of transient ingestion errors")
+    private int maxRetryAttempts = 1;
+
+    @FieldDoc(defaultValue = "10", help = "Period of time in milliseconds to backoff"
+            + " before retry for transient errors")
+    private long retryBackOffTime = 10;
+
+
+    public static ADXSinkConfig load(String yamlFile) throws IOException {
+        ObjectMapper mapper = new ObjectMapper(new YAMLFactory());
+        return mapper.readValue(new File(yamlFile), ADXSinkConfig.class);
+    }
+
+    protected static ADXSinkConfig load(Map<String, Object> config) throws JsonProcessingException {
+        ObjectMapper mapper = new ObjectMapper();
+        return mapper.readValue(mapper.writeValueAsString(config), ADXSinkConfig.class);
+    }
```

Comment:
Please check https://github.com/apache/pulsar/pull/21675 to see how sensitive values are handled there. Another possibility is to use what you have and after this retrieve the secret using the getSecret method on the SinkContext.

Reply:
An alternative is to use [PIP-289](https://github.com/apache/pulsar/blob/master/pip/pip-289.md) support which wouldn't require code changes. However, that's primarily a solution that can be used for existing connectors that don't have proper handling for secrets. The doc is useful to read to get an overview of secrets handling challenges in Pulsar connectors.

Reply:
Thanks for the explanation @lhotari , I checked the sources you provided. Will implement the `IOConfigUtils.loadWithSecrets` way to load the sensitive info.